File: elixir/pe005.exs
defmodule ProjectEuler do
  def pe005(n) do
    lowest_common_multiple(n, n, 1)
  end
   
  defp is_prime?(_, try) when try == 1 do
    true
  end

  defp is_prime?(n, try) when rem(n, try) == 0 do
    false
  end

  defp is_prime?(n, try) do
    is_prime?(n, try - 1)
  end

  defp is_prime?(n) do
    root = Kernel.round :math.sqrt(n)
    is_prime?(n, root)
  end

  defp power_less_than(n, p, max) do
    product = n * p
    if product > max do
      p
    else
      power_less_than(n, product, max)
    end
  end
 
  defp lowest_common_multiple(n, max, product) when n > 1 do
    if is_prime?(n) do
      x = power_less_than(n, 1, max)
      lowest_common_multiple(n - 1, max, product * x)
    else
      lowest_common_multiple(n - 1, max, product)
    end
  end

  defp lowest_common_multiple(_, _, product) do
    product
  end
end

IO.puts(ProjectEuler.pe005(20))

File: haskell/pe005.hs
-- this computes projecteuler.net problem 005
is_prime_try :: Int -> Int -> Bool

is_prime_try _ 1                      = True
is_prime_try n try | n `rem` try == 0 = False
is_prime_try n try                    = is_prime_try n (try - 1)

is_prime :: Int -> Bool
is_prime n = do
  let x = floor (sqrt (fromIntegral n))
  is_prime_try n x

power_less_than :: Int -> Int -> Int -> Int
power_less_than n p maxi =
  do
    let product = n * p
    if product > maxi then
      p
    else
      power_less_than n product maxi

lowest_common_multiple :: Int -> Int -> Int -> Int
lowest_common_multiple n maxi product | n > 1 = 
  if is_prime n then
    do 
      let x = power_less_than n 1 maxi
      lowest_common_multiple (n - 1) maxi (product * x)
  else
    lowest_common_multiple (n - 1) maxi product 

lowest_common_multiple _ _ product = product

pe005 :: Int -> Int
pe005 n = lowest_common_multiple n n 1

main = print(pe005 20)
